src/bin/ion/commands/generate/generator.rs

```diff
@@ -482,22 +489,19 @@ impl<'a, L: Language + 'static> CodeGenerator<'a, L> {
         // TODO: verify the `occurs` value within a field, by default the fields are optional.
         // add current data model node into the data model store
         // verify if the field presence was provided as optional and set the type reference name as optional.
-        let type_name = match field_presence {
-            FieldPresence::Optional => abstract_data_type.fully_qualified_type_ref::<L>().ok_or(
-                invalid_abstract_data_type_raw_error(
-                    "Can not determine fully qualified name for the data model",
-                ),
-            )?,
-            FieldPresence::Required => abstract_data_type.fully_qualified_type_ref::<L>().ok_or(
-                invalid_abstract_data_type_raw_error(
-                    "Can not determine fully qualified name for the data model",
-                ),
-            )?,
-        };
+        let type_name = abstract_data_type.fully_qualified_type_ref::<L>();
 
         self.data_model_store
             .insert(type_name, data_model_node.to_owned());
 
+        // pop out the nested type name from the fully qualified namespace as it has been already added to the type store and to nested types
+        // For sequence type, it would already have popped out the nested type name.
+        if !data_model_node.is_sequence() {
+            // Since the fully qualified name of this generator represents the current fully qualified name,
+            // remove it before generating code for the next ISL type.
+            L::reset_namespace(&mut self.current_type_fully_qualified_name);
```

Comment:
I'm still not sure why we need a `reset_namespace` function. Is it because the data model namespaces are not what we need for generating the code?

Reply:
We do need the data model namespace for generating code. So we have this `add_to_namespace` being called on each ISL type being converted to a data model. But when we are done we modify the namespace back to its original value by resetting the namespace. This makes sure that the next ISl type being converted doesn't get a wrong namespace value.

Reply:
Based on offline discussion, there is still some language-dependent logic used in creating the abstract data model, which is why we have extra namespace manipulation in the generator and why a nested type returns `None` from `fully_qualified_type_name()`. This is something that will need to be fixed before the `generate` command is considered stable/GA, but it doesn't need to be a blocker right now.